src/memory_server.ml

```diff
@@ -278,22 +298,51 @@ let calculate_boot_time_host_free_memory () =
       pages_to_kib (Int64.of_nativeint boot_time_host_free_pages) in
     Int64.mul 1024L boot_time_host_free_kib
 
+let calc_constant_boot_time_host_free_memory constant_count interval =
```

Comment:
See comment 2) above: it assumes no other domain has been started.
This will no longer be true: pv-iommu can take a long time to be initialized, causing squeezed to wait for stable memory, during which time (non-pass-through) VMs can be booted. So I think this calculation needs to be updated to subtract the memory used by domains other than Dom0 too.

The stop condition would be:
* the free memory values you measured in the past N minutes is stable (identical)
* you have not seen a VM start or shutdown event (squeezed must already have a way for watching xenstore for introducedomain/releasedomain). If you see such an event you "reset the clock" and wait N minutes for stabilization again

This could also mean that you are forever in the stabilization phase and never reach a stable point, but on a system that eventually doesn't start/stop more VMs it should reach a stable point.

Reply:
No, we don't allow other VMs to start during this time, because each VM consumes some additional amount of extra host memory, which is badly defined. So we establish the host overhead before starting VMs, otherwise the memory calculations get even more unreliable than they already are. The current change causes xapi to not enable the host until the host overhead has been established.

Reply:
IIUC we have to allow VMs to start, otherwise if you have a very powerful server with lots of memory you will get a 15minute or more delay on boot, where you can't start anything.
This squeezed workaround is needed to avoid that (so that pv-iommu can get initialized in the background, and you can start VMs that don't need pv-iommu in the meantime).
Otherwise we could just wait for pv-iommu to be initialized and tell squeezed to recompute the memory once, but: pv-iommu may be disabled, so you could be waiting forever; we may want to get rid of pv-iommu in general, and this is a workaround until we can do that.

Reply:
.. just read the other comments, I see you intend to make this a blocking API call so that other VMs cannot start until host memory has stabilized, in which case ignore my comment about VMs getting started or stopped, but perhaps we should instead add an assertion/sanity check to count the number of active domains, and bail out if it is more than just Dom0.

Reply:
We have seen issues on large hosts where VMs were started while the PV-IOMMU was busy initialising, even if those VMs were not supposed to "need" the PV-IOMMU. Therefore, we concluded that it is best to not allow to start VMs at all during this time to avoid all issues. Furthermore, I really don't want squeezed to know anything about the PV-IOMMU or have any other specific low level knowledge. In fact, there are other (less understood) reasons why the memory level drops for a while after dom0 has started. This is why this squeezed change is needed, and it is not just a PV-IOMMU workaround.

The PV-IOMMU problem will be addressed separately, by the Hypervisor people, to avoid the long boot times.